src/plugins/stdioproc/README.md

```diff
@@ -0,0 +1,201 @@
+> - followed by a newline
+> - followed by a keyset called `users`
+> - followed by a newline
+> - followed by dynamic text called `user`
+>
+> The names `users` and `user` don't actually appear in the message.
+> They are only used as a reference for the descriptions of the message.
+>
+> The actual message sent could look like this:
+>
+> ```
+> HELLO WORLD
+> kdbOpen 2
+> $key string 3 5
+> joe
+> 12345
+> $end
+> appleseed
+> ```
+
+### Initialization
+
+The parent writes the protocol header to `stdin` of the child.
+
+```
+Parent > Child
+
+ELEKTRA_STDIOPROC INIT v1
+```
+
+The child must respond with an acknowledgement, followed by a `contract` keyset.
+
+```
+Child > Parent
+
+ELEKTRA_STDIOPROC ACK v1
+(name)
+[contract]
+```
+
+Here `(name)` is the module name the child uses.
+This will be used to replace `stdioproc` in the contract of the plugin.
+The `[contract]` keyset must contain the `infos` keys, which will be used replace the ones at the top of this README and the ones in `stdioproc.c`.
+It must also contain `exports/_` keys for every operation that is implemented by the child.
+All keys in `[contract]` should be below `system:/elektra/modules/stdioproc`.
+The parent will rename these keys appropriately.
+
+A child that implements all operations should include these keys (in addition to the relevant `infos` keys):
+
+```
+system:/elektra/modules/stdioproc/exports/open = 1
+system:/elektra/modules/stdioproc/exports/close = 1
+system:/elektra/modules/stdioproc/exports/get = 1
+system:/elektra/modules/stdioproc/exports/set = 1
+```
+
+After this initial handshake, the child should simply wait for further requests from the parent.
+
+> **Note**: Under normal circumstances this handshake will always be followed by an `open` request immediately.
+
+### Operation
+
+When the parent needs the child to process an operation (open, get, ...), it will send a request like this:
+
+```
+Parent > Child
+
+(opname)
+[parent]
+[data]
+```
+
+Here `(opname)` is one of `open`, `get`, `set` or `close` and describes the operation that shall be performed by the child.
+The keyset `[parent]` always consists of a single key, namely the `parentKey` (or `errorKey`) that was passed to the plugin.
+Finally, `[data]` is the keyset that was passed to the plugin.
+The `[data]` keyset is not present in `open` and `close` operations, since those don't receive a `KeySet` in the C API.
+However, in the `open` operation `[data]` is replaced `[config]` which is the `KeySet` returned by `elektraPluginGetConfig` in the C API.
+This is needed, because the child process cannot request the config keyset otherwise.
+
+The child should then perform the requested operation and respond with
+
+```
+Child > Parent
+
+(result)
+[parent]
+[data]
+```
+
+Here `(result)` is one of `success`, `noupdate` and `error`, which correspond to `ELEKTRA_PLUGIN_STATUS_SUCCESS`, `ELEKTRA_PLUGIN_STATUS_NO_UPDATE` and `ELEKTRA_PLUGIN_STATUS_ERROR` respectively.
```

Comment:
Why not send ELEKTRA_PLUGIN_STATUS_SUCCESS and so on?

Reply:
Why is sending a much longer string any better? We already have a handshake, so the `ELEKTRA_PLUGIN_STATUS_` prefix doesn't serve any purpose here.

If you mean sending the integer values of the constants, then this would just make the protocol less human-readable. It wouldn't actually have any benefit, because you'd still need to decode the string e.g. `1` into the integer value.